### PR TITLE
Cluster Configuration 

### PR DESCRIPTION
**Pull request by russjones**
**Purpose**

This PR addresses the issues we've had when it comes to machine and cluster configuration. For further details see https://github.com/gravitational/teleport/issues/1098.  

**Implementation**

Starting with Teleport 2.3, cluster configuration can only be changed by updating `auth_service` in `teleport.yaml`. Cluster configuration consists of cluster name, static provisioning tokens, and authentication preferences.

Teleport will read in the configuration file and create resources of the three items (`services.ClusterName`, `services.StaticTokens`, and `services.AuthPreferences`) and upsert them to the backend. The last Auth Server that starts will be the one that sets cluster configuration. Because cluster configuration has become resources, all Auth Servers will have a consistent answer for cluster configuration in a HA environment.

`services.AuthPreferences` has been updated to also include a `connector_name` field. This field will allow you to add multiple OIDC or SAML resources to your backend and then pick which one you wish to use instead. If no `connector_name` the first one will be selected like in Teleport 2.2.

OIDC connector configuration and Trusted Cluster configuration has been depreciated in Teleport 2.3. For users upgrading from Teleport 2.2, existing OIDC connectors and Trusted Cluster configuration will not be removed, but you will no longer be able to add, remove, or update them. A warning will be printed to the logs indicating that the user should use resources.

The `dynamic_config` has been deprecated  (a warning is printed) and `seed_config` has been completely removed as it was deprecated in Teleport 2.0.

The U2F resource has been completely been removed. Users of Open Source Teleport will see no differences because this connector was specified in file configuration and we will automatically migrate it to `services.AuthPreferences`.

**Related Issues**

Fixes https://github.com/gravitational/teleport/issues/1098

### CODE REVIEW

File: lib/auth/apiserver.go [hunk 1]
```diff
@@ -1589,72 +1596,106 @@ func (s *APIServer) deleteRole(auth ClientI, w http.ResponseWriter, r *http.Requ
...
+
+	return rawMessage(services.GetStaticTokensMarshaler().Marshal(st, services.WithVersion(version)))
+}
+
+type setStaticTokensReq struct {
+	StaticTokens json.RawMessage `json:"static_tokens"`
+}
+
+func (s *APIServer) setStaticTokens(auth ClientI, w http.ResponseWriter, r *http.Request, p httprouter.Params, version string) (interface{}, error) {
+	var req *setStaticTokensReq
```
Comment (klizhentas) [inline lib/auth/apiserver.go]: why pointer to pointer

File: lib/auth/auth_test.go [hunk 2]
```diff
@@ -186,7 +201,13 @@ func (s *AuthSuite) TestTokensCRUD(c *C) {
 
 	// lets use static tokens now
 	roles = teleport.Roles{teleport.RoleProxy}
-	s.a.StaticTokens = append(s.a.StaticTokens, services.ProvisionToken{Token: "static-token-value", Roles: roles, Expires: time.Unix(0, 0)})
+	//s.a.StaticTokens = append(s.a.StaticTokens, services.ProvisionToken{Token: "static-token-value", Roles: roles, Expires: time.Unix(0, 0)})
```
Comment (klizhentas) [inline lib/auth/auth_test.go]: leftover comment

File: lib/config/configuration.go [hunk 3]
```diff
@@ -391,6 +301,15 @@ func ApplyFileConfig(fc *FileConfig, cfg *service.Config) error {
 		cfg.Auth.SSHAddr = *addr
 		cfg.AuthServers = append(cfg.AuthServers, *addr)
 	}
+	// DEPRECATED: Remove the following message in Teleport 2.4
+	if fc.Auth.DynamicConfig != nil {
+		warningMessage := "Dynamic configuration is no longer supported. Refer to " +
+			"Teleport documentation for more details: https://www.example.com."
```
Comment (klizhentas) [inline lib/config/configuration.go]: missing doc link

File: lib/config/configuration.go [hunk 4]
```diff
@@ -399,12 +318,49 @@ func ApplyFileConfig(fc *FileConfig, cfg *service.Config) error {
...
-		cfg.Auth.StaticTokens = append(cfg.Auth.StaticTokens, services.ProvisionToken{Token: tokenValue, Roles: roles, Expires: time.Unix(0, 0)})
+		cfg.ReverseTunnels = append(cfg.ReverseTunnels, tun)
+	}
+	// DEPRECATED: Remove the following message in Teleport 2.4.
+	if len(fc.Auth.TrustedClusters) > 0 {
+		warningMessage := "Configuring Trusted Clusters using file configuration has " +
+			"been deprecated, Trusted Clusters must now be configured using resources. " +
+			"Existing Trusted Cluster relationships will be maintained, but you will " +
+			"not be able to add or remove Trusted Clusters using file configuration " +
+			"anymore. Refer to Teleport documentation for more details: https://www.example.com."
```
Comment (klizhentas) [inline lib/config/configuration.go]: missing link

File: lib/config/configuration.go [hunk 5]
```diff
@@ -399,12 +318,49 @@ func ApplyFileConfig(fc *FileConfig, cfg *service.Config) error {
...
+		}
+		cfg.Auth.Preference = authPreference
+
+		// DEPRECATED: Remove the following message in Teleport 2.4.
+		if oidcConnector != nil {
+			warningMessage := "Configuring OIDC connectors using file configuration is " +
+				"no longer supported, OIDC connectors must be configured using resources. " +
+				"Existing OIDC connectors will not be removed but you will not be able to " +
+				"add, remove, or update them using file configuration. Refer to Teleport " +
+				"documentation for more details: https://www.example.com."
```
Comment (klizhentas) [inline lib/config/configuration.go]: link

File: lib/config/fileconf.go [hunk 6]
```diff
@@ -506,74 +508,112 @@ type TrustedCluster struct {
...
+	if err != nil {
+		return services.ProvisionToken{}, trace.Wrap(err)
 	}
-	roles, err = teleport.ParseRoles(parts[0])
-	return roles, parts[1], trace.Wrap(err)
+
+	return services.ProvisionToken{
+		Token:   parts[1],
+		Roles:   roles,
+		Expires: time.Unix(0, 0),
```
Comment (klizhentas) [inline lib/config/fileconf.go]: .UTC()

File: lib/web/apiserver.go [hunk 7]
```diff
@@ -345,7 +345,13 @@ func (m *Handler) getUserACL(w http.ResponseWriter, r *http.Request, _ httproute
 }
 
 func buildUniversalSecondFactorSettings(authClient auth.ClientI) *client.U2FSettings {
-	universalSecondFactor, err := authClient.GetUniversalSecondFactor()
+	cap, err := authClient.GetAuthPreference()
+	if err != nil {
+		log.Debugf("Unable to get U2F Settings: %v", err)
+		return nil
```
Comment (klizhentas) [inline lib/web/apiserver.go]: why do you return nil here and not `trace.NotFound` and handle on the client

File: lib/web/apiserver.go [hunk 8]
```diff
@@ -356,61 +362,90 @@ func buildUniversalSecondFactorSettings(authClient auth.ClientI) *client.U2FSett
...
+	var err error
+	var oidcConnector services.OIDCConnector
+
+	// if we have a connector name, try and pull that connector back, otherwise
+	// return the first connector
+	connectorName := cap.GetConnectorName()
+	if connectorName != "" {
+		oidcConnector, err = authClient.GetOIDCConnector(connectorName, false)
+		if err != nil {
+			log.Debugf("Unable to find OIDC Connector %v: %v", connectorName, err)
```
Comment (klizhentas) [inline lib/web/apiserver.go]: same here

File: lib/web/apiserver.go [hunk 9]
```diff
@@ -356,61 +362,90 @@ func buildUniversalSecondFactorSettings(authClient auth.ClientI) *client.U2FSett
...
+		if err != nil {
+			// if we have nothing set on the backend, return we have nothing
+			if trace.IsNotFound(err) {
+				return nil
+			}
 
-		log.Debugf("Unable to get OIDC Connectors: %v", err)
-		return nil
-	}
+			log.Debugf("Unable to get OIDC Connectors: %v", err)
```
Comment (klizhentas) [inline lib/web/apiserver.go]: why do you ignore non `NotFound` error

File: lib/web/apiserver.go [hunk 10]
```diff
@@ -356,61 +362,90 @@ func buildUniversalSecondFactorSettings(authClient auth.ClientI) *client.U2FSett
...
+			return nil
+		}
 
-	if len(oidcConnectors) < 1 {
-		log.Debugf("No OIDC Connectors found")
-		return nil
+		if len(oidcConnectors) < 1 {
```
Comment (klizhentas) [inline lib/web/apiserver.go]: == 0

**Comment by russjones** [comment]
@klizhentas Made most of the changes you suggested. Remaining issues are addressed in https://github.com/gravitational/teleport/pull/1188.

File: lib/config/fileconf.go (hunk 6 above)
Comment (russjones) [inline lib/config/fileconf.go]: Looks like they are both equal: https://play.golang.org/p/rZHSZaviB6

That being said, I like adding `.UTC()` as it makes it explicit that we expect all time to be in UTC so I made the change.

**Review by klizhentas — APPROVED** [review]

**Pull request merged**